### ISSUE TITLE
Investigate colour contrast issues with the focus state

### ISSUE THREAD

**Issue by 36degrees** [alphagov/govuk-frontend#1137]
## What

Investigate colour contrast issues with the focus state according to criterion [1.4.11 Non-Text Contrast](https://www.w3.org/WAI/WCAG21/Understanding/non-text-contrast) of WCAG 2.1.

The focus state for an inline control such as a link is to use black text colour against a yellow highlight.

<img width="110" alt="Example of a focussed link" src="https://user-images.githubusercontent.com/121939/51192189-e88e7500-18dd-11e9-8e63-daec27bfd178.png">

The focus state for a form element comprises a yellow outline (`#ffbf47`) around a black border (`#0b0c0c`).

<img width="631" alt="Example of a focussed form input" src="https://user-images.githubusercontent.com/121939/51192183-e62c1b00-18dd-11e9-8275-ea196a9ea6b4.png">

In most cases, these will be used against a white (`#ffffff`) background, but other colour backgrounds should be considered. For example (non-exhaustive):

`#dee0e2` (1.24:1 ❌)

<img width="1046" alt="Example of focussed link against grey in the footer component" src="https://user-images.githubusercontent.com/121939/51192301-2c817a00-18de-11e9-8161-aa1ed4d7ca7f.png">

`#0b0c0c` (11.93:1 ✔️)

<img width="360" alt="Example of a focussed input against black in the header of the Design System" src="https://user-images.githubusercontent.com/121939/51192302-2c817a00-18de-11e9-9d31-377ee2cacc7e.png">

`#005ea5` (4.07:1 ✔️)

<img width="812" alt="Example of a focussed button against blue in the masthead of the Design System" src="https://user-images.githubusercontent.com/121939/51192304-2c817a00-18de-11e9-9412-f8908743e477.png">

When used against a white background, the colour contrast ratio between the yellow and the white is 1.64:1.

## Why

WCAG 2.1 introduces a new success criteria "[1.4.11 Non-Text Contrast](https://www.w3.org/WAI/WCAG21/Understanding/non-text-contrast)" which intends that "any visual effects implemented which are necessary to indicate state, such as whether a component is selected or focused, must also ensure that the information used to identify the control in that state has a minimum 3:1 contrast ratio".

**Comment by nickcolley** [comment]
@dashouse worth a read I think: https://www.w3.org/WAI/WCAG21/Understanding/non-text-contrast

**Comment by stevenaproctor** [comment]
@nickcolley @dashouse This has came up in a DAC audit we had too.

I had a quick look through the colour palette. Those that do not meet the 3:1 for non-text contrast are:

- govuk-border-colour and grey-2
- govuk-focus-colour and yellow
- grey-3
- grey-4
- orange
- light-pink

The same 3:1 rule applies for text contrast too. A link (005ea5) should have a 3:1 colour contrast compared to normal text (0b0c0c). But the contrast is 2.93:1 for a link and 1.94:1 for a visited link. This is part of https://www.w3.org/WAI/WCAG21/Understanding/use-of-color

WCAG 1.4.3 is about foreground and background contrast, which should be 4.5:1. The govuk-link-hover-colour contrast is 3.72:1. See https://www.w3.org/WAI/WCAG21/Understanding/contrast-minimum

**Comment by joelanman** [comment]
@stevenaproctor I might be wrong but I think specifically on links and normal text, the colour only has to contrast if there is no other visual indicator. In the large majority of cases our links have underlines, therefore I don't _think_ they need to also have colour contrast. The 'Use of color' you linked to above says "Color is not used as the only visual means of conveying information".

**Comment by stevenaproctor** [comment]
@joelanman That is exactly what it says and I totally agree. But I think it would be a good to make the contrast better in case the underline is not visible for some reason.

Something from another DAC audit said:

>Another piece of feedback we got from a DAC report was:

> There was an issue with the link highlighting mentioned by one of our low vision analysts where he felt that a clearer indication of when the mouse is placed over a link/button is required.
>
> Solution = Consider improving the mouse hover indication on links and buttons to ensure that users with low vision can clearly detect when the mouse is placed on the elements. It is recommended that the same visual indicator is offered to mouse users as those that navigate using keyboards.

![image](https://user-images.githubusercontent.com/11977962/51252260-43ce6f00-1993-11e9-967a-4cd018e093f9.png)

All the links on W3's site do something similar. The only difference between hover and keyboard focus is the focus outline.

![image](https://user-images.githubusercontent.com/11977962/51252401-a1fb5200-1993-11e9-82c3-95db111dee0b.png)

I am not saying I agree this is the correct thing to do but we should consider it.
